### PR TITLE
Persistent accounts stabilization

### PR DESCRIPTION
**Pull request by s373r**
This PR changes are extracted from
- https://github.com/kamu-data/kamu-cli/pull/650

---

## Description

Related issue: https://github.com/kamu-data/kamu-node/issues/68

### Added
- Added E2E test infrastructure
  - Added necessary components for managed run -- for startup, operations, and shutdown
### Changed
- Hidden arguments and options are excluded from [the CLI reference](resources/cli-reference.md)
### Fixed
- Stabilized startup using connection to databases
  - Added HTTP middleware that wraps each request into a separate transaction 
  - Also added wrapping for some commands, in particular `kamu system generate-token`
  - The structure of services that required lazy access to databases was reorganized:
     - Extracted `PredefinedAccountsRegistrator` & `DatasetOwnershipServiceInMemoryStateInitializer` 
- Fixed potential crash when attempting to rollback a transaction if the connection fails to establish

## Checklist before requesting a review

- [x] Unit and integration tests added, ✅
  - E2E tests added
  
- [x] Compatibility:
    
  - [x] Network APIs: ✅
    
  - [x] Workspace layout and metadata: ✅ 
    
  - [x] Configuration: ✅
    
  - [x] Container images: ✅
- [x] Documentation:
    
  - [x] [Changelog](./CHANGELOG.md): ✅
    
  - [x] [Public documentation](https://github.com/kamu-data/kamu-docs/): ❌, not needed
  
- [x] Downstream effects:
    
  - [x] [kamu-node](https://github.com/kamu-data/kamu-node): ❌, will be added later

### CODE REVIEW

File: resources/cli-reference.md [hunk 1]
```diff
@@ -306,8 +307,8 @@ Initialize an empty workspace in the current directory
 
 * `--exists-ok` — Don't return an error if workspace already exists
 * `--pull-images` — Only pull container images and exit
-* `--list-only` — List image names instead of pulling
-* `--multi-tenant` — Initialize a workspace for multiple tenants
+* `--list-only` — List image names instead of pulling (hidden)
+* `--multi-tenant` — Initialize a workspace for multiple tenants (hidden)
```
Comment (zaychenko-sergei) [inline resources/cli-reference.md:311]: Offtopic: I wonder if content for hidden switches should be generated into the reference.

File: src/app/cli/src/app.rs [hunk 2]
```diff
@@ -135,19 +135,38 @@ pub async fn run(
...
-                Err(CLIError::usage_error_from(WorkspaceUpgradeRequired))
-            } else if current_account.is_explicit() && !is_multi_tenant_workspace {
-                Err(CLIError::usage_error_from(NotInMultiTenantWorkspace))
-            } else {
-                command.run().await
+    // TODO: Extend Command trait with the corresponding property
+    let need_to_wrap_with_transaction = matches!(
+        matches.subcommand(),
+        Some(("login" | "add" | "delete" | "pull", _))
+    );
```
Comment (zaychenko-sergei) [inline src/app/cli/src/app.rs:142]: I think there is nothing to wait for, do this extension

File: src/domain/accounts/services/src/authentication_service_impl.rs [hunk 3]
```diff
@@ -65,47 +64,34 @@ impl AuthenticationServiceImpl {
...
 
     fn resolve_authentication_provider(
         &self,
         login_method: &str,
-    ) -> Result<Arc<dyn AuthenticationProvider>, UnsupportedLoginMethodError> {
-        match self.authentication_providers_by_method.get(login_method) {
+        transactional_catalog: &TransactionalCatalog,
+    ) -> Result<Arc<dyn AuthenticationProvider>, LoginError> {
+        let authentication_providers =
+            self.get_authentication_providers_by_method(transactional_catalog);
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/authentication_service_impl.rs:87]: It's not using `login_method`, so why calling it like this.

File: src/domain/accounts/services/src/authentication_service_impl.rs [hunk 4]
```diff
@@ -198,182 +187,353 @@ impl AuthenticationServiceImpl {
...
+                    .keys()
+                    .copied()
+                    .collect();
+
+                // Stable order for tests
+                login_methods.sort_unstable();
+
+                Ok(login_methods)
+            })
+            .await
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/authentication_service_impl.rs:253]: I don't fully understand why this requires a transaction

File: src/domain/accounts/services/src/login_password_auth_provider.rs [hunk 5]
```diff
@@ -88,10 +88,14 @@ impl AuthenticationProvider for LoginPasswordAuthProvider {
         &self,
         login_credentials_json: String,
     ) -> Result<ProviderLoginResponse, ProviderLoginError> {
+        dbg!("@@@ login 1", &login_credentials_json);
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/login_password_auth_provider.rs:91]: If we want some logs to remain, use `tracing`

**Comment by zaychenko-sergei** [comment]
What about Github Actions support for new testing type?

File: src/domain/accounts/services/src/authentication_service_impl.rs [hunk 6]
```diff
@@ -32,22 +33,20 @@ use serde::{Deserialize, Serialize};
...
     predefined_accounts_config: Arc<PredefinedAccountsConfig>,
     time_source: Arc<dyn SystemTimeSource>,
     encoding_key: EncodingKey,
     decoding_key: DecodingKey,
-    authentication_providers_by_method: HashMap<&'static str, Arc<dyn AuthenticationProvider>>,
-    login_password_auth_provider: Option<Arc<LoginPasswordAuthProvider>>,
-    account_repository: Arc<dyn AccountRepository>,
+    transaction_runner: Arc<DatabaseTransactionRunner>,
     state: tokio::sync::Mutex<State>,
 }
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/authentication_service_impl.rs:51]: A few things should be changed:

1) We are not strictly required to do lazy registration of predefined accounts. This could be moved to a separate "initializer" service that is only invoked from `app.py` (and it's `node` equivalent). The initialization code will be contained in it's own transaction.

2) If suggestion above works, we can get rid of `State` and `Singleton` status

3) I think at the end of refactoring, `AuthenticationServiceImpl` should not be responsible for controlling the scope of transaction. In case of web queries, there is middleware. In case of commands, there is a transaction in `app.py` around execution of commands.

File: resources/cli-reference.md (hunk 1 above)
Comment (s373r) [reply resources/cli-reference.md:311]: I've been thinking about that, too -- I think it's more of a no than a yes

File: src/domain/accounts/services/src/login_password_auth_provider.rs (hunk 5 above)
Comment (s373r) [reply src/domain/accounts/services/src/login_password_auth_provider.rs:91]: oh, it wasn't supposed to be commited...

File: src/e2e/app/cli/postgres/tests/tests/test_auth.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                "#
+            ),
+            host = db.get_host(),
+            user = db.get_username(),
+            password = db.get_username(), // It's intended: password is same as user for tests
+            database = db.get_database().unwrap(),
+        )
+        .as_str(),
+    )
+    .await;
```
Comment (zaychenko-sergei) [inline src/e2e/app/cli/postgres/tests/tests/test_auth.rs:42]: Configuration/workspace code repeats, isolate it into harness

File: src/e2e/app/cli/postgres/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+          AND state = 'idle'
+          AND pid <> pg_backend_pid()   -- exclude the current connection
+          AND query = 'COMMIT'          -- last executed query
+          AND wait_event = 'ClientRead' -- awaiting new data
+        "#,
+        database
+    )
+    .fetch_one(&pg_pool)
+    .await
+    .unwrap();
```
Comment (zaychenko-sergei) [inline src/e2e/app/cli/postgres/src/lib.rs:54]: I don't understand why is this necessary.

File: src/domain/accounts/services/src/authentication_service_impl.rs (hunk 6 above)
Comment (s373r) [reply src/domain/accounts/services/src/authentication_service_impl.rs]: > This could be moved to a separate "initializer" service

Added `PredefinedAccountsRegistrator` -- done 

> If suggestion above works, we can get rid of State and Singleton status
> ...
> I think at the end of refactoring, AuthenticationServiceImpl should not be responsible for controlling the scope of transaction`

Done

File: src/domain/accounts/services/src/authentication_service_impl.rs (hunk 4 above)
Comment (s373r) [reply src/domain/accounts/services/src/authentication_service_impl.rs]: `AuthenticationServiceImpl` no longer manages transactions

---

> I don't fully understand why this requires a transaction

But to answer the question: it was a necessity from a DI perspective

File: src/e2e/app/cli/postgres/tests/tests/test_auth.rs (hunk 7 above)
Comment (s373r) [reply src/e2e/app/cli/postgres/tests/tests/test_auth.rs]: Added `kamu_cli_e2e_test!()` macro

File: src/app/cli/src/app.rs (hunk 2 above)
Comment (s373r) [reply src/app/cli/src/app.rs]: Unfortunately, since we need to know if there will be a transaction (!) even before a command is created, I had to do it with a separate function `cli_commands::command_needs_transaction()`

File: deny.toml [hunk 9]
```diff
@@ -86,5 +86,5 @@ unsound = "warn"
 yanked = "deny"
 notice = "warn"
 ignore = [
+    "RUSTSEC-2023-0071"
```
Comment (zaychenko-sergei) [inline deny.toml:89]: A comment why would be nice

File: src/app/cli/src/app.rs [hunk 10]
```diff
@@ -135,22 +142,64 @@ pub async fn run(
...
+                .get_one::<DatasetOwnershipServiceInMemoryStateInitializer>()
+                .map_err(CLIError::critical)?;
+
+            initializer
+                .eager_initialization()
+                .await
+                .map_err(CLIError::critical)
+        })
+        .await?;
+
```
Comment (zaychenko-sergei) [inline src/app/cli/src/app.rs:171]: Minor: let's make method from these

File: src/domain/accounts/services/src/predefined_accounts_registrator.rs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            })?;
+
+        if account_config.provider == PROVIDER_PASSWORD
+            && let Some(login_password_auth_provider) =
+                self.maybe_login_password_auth_provider.as_ref()
+        {
+            login_password_auth_provider
+                .save_password(&account.account_name, account_config.get_password())
+                .await?;
+        }
```
Comment (zaychenko-sergei) [inline src/domain/accounts/services/src/predefined_accounts_registrator.rs:91]: Perhaps it should panic, if this dependency is undefined. Or we could always demand it instead of Option

File: deny.toml (hunk 9 above)
Comment (s373r) [reply deny.toml]: It was a backport of a change from [another PR](https://github.com/kamu-data/kamu-cli/pull/665) (merged) but I've added context as to why the change was made

File: src/app/cli/src/app.rs (hunk 10 above)
Comment (s373r) [reply src/app/cli/src/app.rs]: Done

File: src/domain/accounts/services/src/predefined_accounts_registrator.rs (hunk 11 above)
Comment (s373r) [reply src/domain/accounts/services/src/predefined_accounts_registrator.rs:91]: `LoginPasswordAuthProvider` is now a required dependency -- done

**Review by zaychenko-sergei — APPROVED** [review]

File: src/app/cli/src/cli_commands.rs [hunk 12]
```diff
@@ -580,6 +583,22 @@ pub fn get_command(
...
+    match arg_matches.subcommand() {
+        Some(("system", system_matches)) => match system_matches.subcommand() {
+            Some(("generate-token", _)) => Ok(true),
+            Some(_) => Ok(false),
+            None => Err(CommandInterpretationFailed.into()),
+        },
+        Some(_) => Ok(false),
+        None => Err(CommandInterpretationFailed.into()),
+    }
+}
```
Comment (sergiimk) [inline src/app/cli/src/cli_commands.rs:596]: This seems super fragile. What stops us from always having some transaction runner in the catalog (real DB or noop impl) and always run commands in transactional scope?

Comment (s373r) [reply src/app/cli/src/cli_commands.rs:596]: we have commands that take a long time to execute -- e.g.  API server

can't do that because the commit will never happen

-- 

The key point here is not the presence of the component itself in the Catalog (they are already present, after all), but the launch of the transaction

Comment (s373r) [reply src/app/cli/src/cli_commands.rs:596]: and yes, I was bitterly disappointed to discover that we need to know if a transaction is required BEFORE creating a Command

File: src/e2e/app/cli/common/src/e2e_hartness.rs [hunk 13]
```diff
@@ -0,0 +1,174 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (s373r) [reply src/e2e/app/cli/common/src/e2e_hartness.rs:1]: Fixed via 
- https://github.com/kamu-data/kamu-cli/pull/667

File: src/utils/database-common/src/transactions/db_transaction_manager.rs [hunk 14]
```diff
@@ -102,8 +135,9 @@ impl TransactionRef {
     pub fn into_maybe_transaction<DB: sqlx::Database>(
         self,
     ) -> Option<sqlx::Transaction<'static, DB>> {
-        let m = Arc::try_unwrap(self.inner).unwrap().into_inner();
-        m.maybe_transaction
+        Arc::try_unwrap(self.inner)
+            .ok()
+            .and_then(|m| m.into_inner().maybe_transaction)
```
Comment (sergiimk) [inline src/utils/database-common/src/transactions/db_transaction_manager.rs:140]: This seems very dangerous - we `unwrap()` the result of extracting from `Arc` to ensure that `into_maybe_transaction` is called ONLY when there are no components holding it as a reference somewhere - i.e. that transaction is not "stuck" in some singleton. Without the `unwrap()` this situation will go unnoticed.

Or am I reading this wrong?

Comment (s373r) [reply src/utils/database-common/src/transactions/db_transaction_manager.rs:140]: You read that correctly -- yes, that change needs to be removed
